### PR TITLE
feat: Calendar `onSelect` support `info.source` param

### LINKED ISSUE #39801: 5.1.0 calendar组件的 onSelect、onChange 似乎有错，和认知中表现不一致
### Reproduction link

[Edit on CodeSandbox](https://codesandbox.io/s/qia-pian-mo-shi-antd-5-1-0-forked-pwsrhn)

### Steps to reproduce

点击任何日期  或者 切换月份，然后查看 console的输出

### What is expected?

onSelect 方法我理解中，是当我选中某个 day后才会触发，但是我切换月份（比如改成1月，再改成3月）、年份 也会触发，
那么请问，我要怎么记录用户选择的日期呢？  
假如用户选择了【2022年1月1日】， 这时候，我会在 onSelect 里读取值 并存入状态， 
这时候用户 切换到了2月（此时面板变成了2月，但用户不去点击某个日） ， 这时候onSelect又触发了， 状态就变了，

我只是想要记录，用户真正点击的日期， 用户切换月份，年份，我不关心， 我只关心用户点击了某个日后，我可以拿到 年月日

我尝试了 onSelect onChange 都不行， 而这个组件也并没有其他 回调 可以拿到 我想要的值

简单总结，有没有某个回调，只有当用户点击具体的 day （日、天）  才会触发

不知道我的表述 是否明朗，  谢谢 

### What is actually happening?

如上

| Environment | Info |
| --- | --- |
| antd | 5.1.0 |
| React | 18.2 |
| System | mac |
| Browser | chrome最新版 |

### LINKED ISSUE #40407: Calendar:  add option to make the date selectable ONLY when you click the specific DAY
### What problem does this feature solve?

When I need to go to a specific date in a year and month different from the current selected date, it will require me to select the date 3 times just to get there. First I have to select the year, then select the month and only then select the day which is exhaustive and unnecessary IMO as it gives a really bad UX.

### What does the proposed API look like?

selectable: 'date' | 'month' | 'year'

### PR DESCRIPTION
**Pull request by zombieJ**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 💡 Background and solution

resolve #39801
resolve #40407
close #41553
close #42412

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |    Calendar `onSelect` support `info.source` param to help get select source.       |
| 🇨🇳 Chinese |    Calendar `onSelect` 支持 `info.source` 参数以获取选择来源。      |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

---

### <samp>🤖 Generated by Copilot at 80c155a</samp>

Added a new feature to the `Calendar` component that allows users to access the source of the date selection, such as 'date', 'keyboard', or 'customize'. Enhanced the `onSelect` and `onValueChange` props with a second parameter `info` of type `SelectInfo` to pass the source information. Updated the tests and the documentation accordingly.

### <samp>🤖 Generated by Copilot at 80c155a</samp>

*  Add a second parameter to the `onSelect` prop of the `Calendar` component to indicate the source of the date selection ([link](https://github.com/ant-design/ant-design/pull/42432/files?diff=unified&w=0#diff-841e97dfb23daba72a34ee15810c8b36da6464eba16e120739ec6c9636f94fb2L71-R75), [link](https://github.com/ant-design/ant-design/pull/42432/files?diff=unified&w=0#diff-841e97dfb23daba72a34ee15810c8b36da6464eba16e120739ec6c9636f94fb2L201-R208), [link](https://github.com/ant-design/ant-design/pull/42432/files?diff=unified&w=0#diff-841e97dfb23daba72a34ee15810c8b36da6464eba16e120739ec6c9636f94fb2L313-R319), [link](https://github.com/ant-design/ant-design/pull/42432/files?diff=unified&w=0#diff-841e97dfb23daba72a34ee15810c8b36da6464eba16e120739ec6c9636f94fb2L335-R343), [link](https://github.com/ant-design/ant-design/pull/42432/files?diff=unified&w=0#diff-fbade99cf5fe70d3001d5570bf707ff05eaa054eb251848d924e7f65e789b807L8-R8), [link](https://github.com/ant-design/ant-design/pull/42432/files?diff=unified&w=0#diff-fbade99cf5fe70d3001d5570bf707ff05eaa054eb251848d924e7f65e789b807L150-R150), [link](https://github.com/ant-design/ant-design/pull/42432/files?diff=unified&w=0#diff-fbade99cf5fe70d3001d5570bf707ff05eaa054eb251848d924e7f65e789b807L176-R188), [link](https://github.com/ant-design/ant-design/pull/42432/files?diff=unified&w=0#diff-8fab8ed47a3daf0635a3eb887b1fd4fdb984ec4daba5af51505fea4a5d87051bL58-R58), [link](https://github.com/ant-design/ant-design/pull/42432/files?diff=unified&w=0#diff-8fab8ed47a3daf0635a3eb887b1fd4fdb984ec4daba5af51505fea4a5d87051bR77-R90), [link](https://github.com/ant-design/ant-design/pull/42432/files?diff=unified&w=0#diff-58e290dfe67aeb90a9d97d4e3e8a790c9aa964113593229b7f2637b66f0c8a50L63-R63), [link](https://github.com/ant-design/ant-design/pull/42432/files?diff=unified&w=0#diff-58e290dfe67aeb90a9d97d4e3e8a790c9aa964113593229b7f2637b66f0c8a50R82-R95))
*  Update the test cases for the `Calendar` component to pass the second parameter to the `onSelect` prop ([link](https://github.com/ant-design/ant-design/pull/42432/files?diff=unified&w=0#diff-0a5821c1b32fdf9bf1d3bb55702937aead9206b6762df0497ff7b030f405a21aL79-R79), [link](https://github.com/ant-design/ant-design/pull/42432/files?diff=unified&w=0#diff-0a5821c1b32fdf9bf1d3bb55702937aead9206b6762df0497ff7b030f405a21aL273-R273), [link](https://github.com/ant-design/ant-design/pull/42432/files?diff=unified&w=0#diff-0a5821c1b32fdf9bf1d3bb55702937aead9206b6762df0497ff7b030f405a21aL282-R282), [link](https://github.com/ant-design/ant-design/pull/42432/files?diff=unified&w=0#diff-0a5821c1b32fdf9bf1d3bb55702937aead9206b6762df0497ff7b030f405a21aL305-R305), [link](https://github.com/ant-design/ant-design/pull/42432/files?diff=unified&w=0#diff-0a5821c1b32fdf9bf1d3bb55702937aead9206b6762df0497ff7b030f405a21aL327-R327), [link](https://github.com/ant-design/ant-design/pull/42432/files?diff=unified&w=0#diff-1adf6928f2944623fff32ce60656db37caf630f5931f17db876c1308f0ffa84aR1-R95))
*  Reorder some imports in `generateCalendar.tsx` and `Header.tsx` to follow the alphabetical order ([link](https://github.com/ant-design/ant-design/pull/42432/files?diff=unified&w=0#diff-841e97dfb23daba72a34ee15810c8b36da6464eba16e120739ec6c9636f94fb2L3-L4), [link](https://github.com/ant-design/ant-design/pull/42432/files?diff=unified&w=0#diff-841e97dfb23daba72a34ee15810c8b36da6464eba16e120739ec6c9636f94fb2R8-R9), [link](https://github.com/ant-design/ant-design/pull/42432/files?diff=unified&w=0#diff-841e97dfb23daba72a34ee15810c8b36da6464eba16e120739ec6c9636f94fb2L17-R18))

### CODE REVIEW

**Review by li-jia-nan — APPROVED** [review]

**Review by kiner-tang — APPROVED** [review]

**Comment by bombillazo** [comment]
Hello, when is this new feature going to be available?

File: components/calendar/generateCalendar.tsx [hunk 1]
```diff
@@ -332,7 +338,9 @@ function generateCalendar<DateType>(generateConfig: GenerateConfig<DateType>) {
           locale={contextLocale?.lang}
           generateConfig={generateConfig}
           cellRender={mergedCellRender}
-          onSelect={onInternalSelect}
+          onSelect={(nextDate) => {
+            onInternalSelect(nextDate, 'date');
+          }}
```
Comment (bombillazo) [inline components/calendar/generateCalendar.tsx:343]: I believe this should be like this:
```tsx
onSelect={(nextDate) => {
  onInternalSelect(nextDate, panelMode === 'date' ? 'date' : 'month');
}}
```

**Review by bombillazo — COMMENTED** [review]
There might be a fix needed for this.

File: components/calendar/generateCalendar.tsx (hunk 1 above)
Comment (li-jia-nan) [reply components/calendar/generateCalendar.tsx:343]: Welcome PR !

Comment (yoyo837) [reply components/calendar/generateCalendar.tsx:343]: @bombillazo 
How about:
```js
onSelect={(nextDate) => {
  onInternalSelect(nextDate, panelMode);
}}
```

Comment (li-jia-nan) [reply components/calendar/generateCalendar.tsx:343]: - #42459

Comment (yoyo837) [reply components/calendar/generateCalendar.tsx:343]: > * [Fix calendar internal select source type #42459](https://github.com/ant-design/ant-design/pull/42459)

Thanks for the reference.

Comment (li-jia-nan) [reply components/calendar/generateCalendar.tsx:343]: Good afternoon, you are welcome

**Pull request merged**